File: cpp/131-PalindromePartitioning.cpp
class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> res;
        vector<string> curr;
        if (s.size() == 0) {
            return res;
        }
        generate(res, curr, s, 0);
        return res;
    }

    void generate(vector<vector<string>>& res, vector<string>& curr, string& s, int start) {
        if (start == s.size()) {
            res.push_back(curr);
            return;
        }
        for (int i = start; i < s.size(); i++) {
            if (isPal(s, start, i)) {
                curr.push_back(s.substr(start, i - start + 1));
                generate(res, curr, s, i + 1);
                curr.pop_back();
            }
        }
    }

    bool isPal(string s, int i, int j) {
        while (i <= j) {
            if (s[i] != s[j])
                return false;
            i++;
            j--;
        }
        return true;
    }
};

File: Java/131-PalindromePartitioning.java
public class Solution {
    public List<List<String>> partition(String s) {
        List<List<String>> res = new ArrayList<>();
        List<String> tmp = new ArrayList<>();
        if (s.length() == 0){
            return res;
        }
        trackback(s, 0, tmp, res);
        return res;
    }
    
    private void trackback(String s, int pos, List<String> tmp, List<List<String>> res){
        if (pos == s.length())  
            res.add(new ArrayList<String>(tmp));
        else{
            for (int i = pos; i < s.length(); i++){
                if (isPal(s, pos, i)){
                    tmp.add(s.substring(pos, i+1));
                    trackback(s, i+1, tmp, res);
                    tmp.remove(tmp.size()-1);
                }
            }
        }
    }
    
    private boolean isPal(String s, int low, int high){
        while (low < high){
            if (s.charAt(low) != s.charAt(high)){
                return false;
            }
            low++;
            high--;
        }
        return true;
    }
}